modular_skyrat/modules/time_clock/code/console.dm

```diff
@@ -0,0 +1,248 @@
+	update_static_data_for_all_viewers()
+	to_chat(user, span_boldwarning("Before clocking out, please return any piece of job gear that is important or limted to your workplace."))
+
+	if(important_job_check())
+		if(SSsecurity_level.get_current_level_as_number() >= job_check_alert_level)
+			to_chat(user, span_boldwarning("You are unable to clock out at the current alert level."))
+			eject_inserted_id(user)
+			return FALSE
+
+		if(tgui_alert(user, "You are a member of security and/or command, make sure that you ahelp before clocking out. If you decide to clock back in later, you will need to go to the Head of Personnel. Do you wish to continue?", "[src]", list("Yes", "No")) != "Yes")
+			eject_inserted_id(user)
+			return FALSE
+
+	playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, FALSE)
+	return TRUE
+
+/obj/machinery/time_clock/AltClick(mob/user)
+	. = ..()
+	if(!Adjacent(user))
+		to_chat(user, span_warning("You are out of range of the [src]!"))
+		return FALSE
+
+	if(!eject_inserted_id(user))
+		return FALSE
+
+///Ejects the ID stored inside of the parent machine, if there is one.
+/obj/machinery/time_clock/proc/eject_inserted_id(mob/recepient)
+	if(!inserted_id || !recepient)
+		return FALSE
+
+	inserted_id.forceMove(drop_location())
+	recepient.put_in_hands(inserted_id)
+
+	inserted_id = FALSE
+	icon_state = "timeclock"
+	update_static_data_for_all_viewers()
+	playsound(src, 'sound/machines/terminal_eject.ogg', 50, FALSE)
+
+	return TRUE
+
+///Clocks out the currently inserted ID Card
+/obj/machinery/time_clock/proc/clock_out()
+	if(!inserted_id)
+		return FALSE
+
+	var/datum/component/off_duty_timer/timer_component = inserted_id.AddComponent(/datum/component/off_duty_timer, CLOCK_IN_COOLDOWN)
+	if(important_job_check())
+		timer_component.hop_locked = TRUE
+		log_admin("[inserted_id.registered_name] clocked out as a head of staff and/or command")
+
+	var/current_assignment = inserted_id.assignment
+	var/datum/id_trim/job/current_trim = inserted_id.trim
+	var/datum/job/clocked_out_job = current_trim.job
+	clocked_out_job.current_positions--
+
+	radio.talk_into(src, "[inserted_id.registered_name], [current_assignment] has gone off-duty.", announcement_channel)
+	update_static_data_for_all_viewers()
+
+	SSid_access.apply_trim_to_card(inserted_id, target_trim, FALSE)
+	inserted_id.assignment = "Off-Duty " + current_assignment
+	inserted_id.update_label()
+
+	GLOB.manifest.modify(inserted_id.registered_name, inserted_id.assignment, inserted_id.get_trim_assignment())
+	return TRUE
+
+///Clocks the currently inserted ID Card back in
+/obj/machinery/time_clock/proc/clock_in()
+	if(!inserted_id)
+		return FALSE
+
+	if(id_cooldown_check())
+		return FALSE
+
+	var/datum/component/off_duty_timer/id_component = inserted_id.GetComponent(/datum/component/off_duty_timer)
+
+	var/datum/job/clocked_in_job = id_component.stored_trim.job
+	if(clocked_in_job.total_positions <= clocked_in_job.current_positions)
+		return FALSE
```

Comment:
That `id_component` can be null, your `id_cooldown_check()` returns false if it doesn't exist on the ID card.

This can therefore potentially cause runtimes.

Reply:
No, it can't be. The `id_cooldown_check` proc returns false if the ID lacks a component, and if that proc returns false, the `clock_in` proc returns false.

Reply:
Perhaps, but if that behavior changes in the future, then it could theoretically runtime. Null checks don't cost much, especially not in code as cold as this. It's just future-proofing, and idiot-proofing your code.

Reply:
I don't think this is a case where there that would apply, but I will do it regardless.